langroid/agent/chat_agent.py

```diff
@@ -581,6 +655,108 @@ def enable_message(
+        """
+        Returns the current set of enabled requests for inference and tools configs.
+        Used for restoring setings overriden by `set_output_format`.
+        """
+        return (
+            self.enabled_requests_for_inference,
+            self.config.use_functions_api,
+            self.config.use_tools,
+        )
+
+    @property
+    def all_llm_tools_handled(self) -> set[str]:
+        """
+        All handled tools; if `output_format` is a `ToolMessage` and
+        the `enable_output_format_handling` config option is enabled,
+        adds `output_format`.
+        """
+        handled = self.llm_tools_handled
+
+        if (
+            self.output_format is not None
+            and issubclass(self.output_format, ToolMessage)
+            and self.config.enable_output_format_handling
+        ):
+            return handled.union({self.output_format.default_value("request")})
+
+        return handled
+
+    @property
+    def all_llm_tools_known(self) -> set[str]:
+        """All known tools; we include `output_format` if it is a `ToolMessage`."""
+        known = self.llm_tools_known
+
+        if self.output_format is not None and issubclass(
+            self.output_format, ToolMessage
+        ):
+            return known.union({self.output_format.default_value("request")})
+
+        return known
+
+    def set_output_format(
+        self, output_type: Optional[type], force_tools: Optional[bool] = None
+    ) -> None:
+        """
+        Sets `output_format` to `output_type` and, if `force_tools` is enabled,
+        switches to the native Langroid tools mechanism to ensure that no tool
+        calls not of `output_type` are generated. By default, `force_tools`
+        follows the `use_tools_on_output_format` parameter in the config.
+
+        If `output_type` is None, restores to the state prior to setting
+        `output_format`.
+        """
+        if output_type is None:
+            self.output_format = None
+            (
+                requests_for_inference,
+                use_functions_api,
+                use_tools,
+            ) = self.saved_requests_and_tool_setings
+            self.config = copy.copy(self.config)
+            self.enabled_requests_for_inference = requests_for_inference
+            self.config.use_functions_api = use_functions_api
+            self.config.use_tools = use_tools
+        else:
+            if force_tools is None:
+                force_tools = self.config.use_tools_on_output_format
+
+            if not any(
+                (isclass(output_type) and issubclass(output_type, t))
+                for t in [ToolMessage, BaseModel]
+            ):
+                output_type = get_pydantic_wrapper(output_type)
+
+            if self.output_format is None and force_tools:
+                self.saved_requests_and_tool_setings = (
+                    self._requests_and_tool_settings()
+                )
+
+            self.output_format = output_type
```

Comment:
when we get to this point are we assuming `self.output_format` is `None`?
Otherwise, we'd be over-writing it, but maybe that is what is intended anyway.

Reply:
We are overriding it; we only consider whether it is `None` when we decide whether to save/restore tools settings.